pipeline-model-declarative-agent/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/DeclarativeAgentDescriptor.java

```diff
@@ -0,0 +1,194 @@
+     *
+     * @return An instantiated {@link DeclarativeAgent}
+     * @throws Exception
+     */
+    public DeclarativeAgent newInstance() throws Exception {
+        return clazz.newInstance();
+    }
+
+    public static ExtensionList<DeclarativeAgentDescriptor> all() {
+        return ExtensionList.lookup(DeclarativeAgentDescriptor.class);
+    }
+
+    /**
+     * Get a map of name-to-{@link DescribableModel} of all known/registered descriptors.
+     *
+     * @return A map of name-to-{@link DescribableModel}s
+     */
+    public static Map<String,DescribableModel> getDescribableModels() {
+        Map<String,DescribableModel> models = new HashMap<>();
+
+        for (DeclarativeAgentDescriptor d : getOrderedDescriptors()) {
+            for (String s : SymbolLookup.getSymbolValue(d)) {
+                models.put(s, new DescribableModel<>(d.clazz));
+            }
+        }
+
+        return models;
+    }
+
+    /**
+     * Get the map of the subset of descriptors with no additional arguments - i.e., "none" and "any".
+     * @return A map of descriptors with no arguments.
+     */
+    public static Map<String,DescribableModel> zeroArgModels() {
+        Map<String,DescribableModel> models = new HashMap<>();
+
+        for (Map.Entry<String,DescribableModel> entry : getDescribableModels().entrySet()) {
+            if (entry.getValue().getParameters().isEmpty()) {
+                models.put(entry.getKey(), entry.getValue());
+            }
+        }
+
+        return models;
+    }
+
+    /**
+     * An ordered (by Extension ordinal, in descending order) list of all descriptors.
+     * @return A list of descriptors.
+     */
+    public static List<DeclarativeAgentDescriptor> getOrderedDescriptors() {
+        List<DeclarativeAgentDescriptor> orderedDescriptors = new ArrayList<>();
+
+        List<ExtensionComponent<DeclarativeAgentDescriptor>> extensionComponents = new ArrayList<>(all().getComponents());
+        Collections.sort(extensionComponents);
+
+        for (ExtensionComponent<DeclarativeAgentDescriptor> extensionComponent: extensionComponents) {
+            orderedDescriptors.add(extensionComponent.getInstance());
+        }
+
+        return orderedDescriptors;
+    }
+
+    /**
+     * An ordered list of all descriptor names.
+     *
+     * @return A list of names
+     */
+    public static List<String> getOrderedNames() {
+        List<String> orderedNames = new ArrayList<>();
+
+        for (DeclarativeAgentDescriptor d : getOrderedDescriptors()) {
+            orderedNames.addAll(SymbolLookup.getSymbolValue(d));
+        }
+
+        return orderedNames;
+    }
+
+    /**
+     * Get the descriptor for a given name or null if not found.
+     *
+     * @param name The name for the descriptor to look up
+     * @return The corresponding descriptor or null if not found.
+     */
+    @Whitelisted
+    public static @Nullable DeclarativeAgentDescriptor byName(@Nonnull String name) {
```

Comment:
Okiedokie, so I'll remove `@Nullable`, since I've just been cargo-culting it. =)